kiwi/xsl/convert72to73.xsl

```diff
@@ -0,0 +1,54 @@
+<?xml version="1.0" encoding="utf-8"?>
+<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
+<xsl:output method="xml"
+        indent="yes" omit-xml-declaration="no" encoding="utf-8"/>
+
+<!-- default rule -->
+<xsl:template match="*" mode="conv72to73">
+    <xsl:copy>
+        <xsl:copy-of select="@*"/>
+        <xsl:apply-templates mode="conv72to73"/>
+    </xsl:copy>
+</xsl:template>
+
+<!-- version update -->
+<para xmlns="http://docbook.org/ns/docbook">
+    Changed attribute <tag class="attribute">schemaversion</tag>
+    to <tag class="attribute">schemaversion</tag> from
+    <literal>7.2</literal> to <literal>7.3</literal>.
+</para>
+<xsl:template match="image" mode="conv72to73">
+    <xsl:choose>
+        <!-- nothing to do if already at 7.3 -->
+        <xsl:when test="@schemaversion > 7.2">
+            <xsl:copy-of select="/"/>
+        </xsl:when>
+        <!-- otherwise apply templates -->
+        <xsl:otherwise>
+            <image schemaversion="7.3">
+                <xsl:copy-of select="@*[local-name() != 'schemaversion']"/>
+                <xsl:apply-templates  mode="conv72to73"/>
+            </image>
+        </xsl:otherwise>
+    </xsl:choose>
+</xsl:template>
+
+<!-- change vmx type to oem type with oem-resize set to false-->
+<xsl:template match="type" mode="conv72to73">
+    <xsl:choose>
+        <xsl:when test="@image='vmx'">
+            <type image="oem">
+                <xsl:copy-of select="@*[local-name() != 'image']"/>
+                <oemconfig>
+                    <oem-resize>false</oem-resize>
+                </oemconfig>
+                <xsl:apply-templates  mode="conv72to73"/>
+            </type>
+        </xsl:when>
+        <xsl:otherwise>
+            <xsl:copy-of select="."/>
+        </xsl:otherwise>
+    </xsl:choose>
+</xsl:template>
```

Comment:
Here it would be great to get some help. If an oemconfig section existed before this stylesheet produces another one and the XML becomes invalid

Reply:
Just for the record, I've suggested this snippet:

```xml
<xsl:template match="type[@image = 'vmx'][not(oemconfig)]" mode="conv72to73">
      <type image="oem">
         <xsl:copy-of select="@*[local-name() != 'image']"/>
         <oemconfig>
            <oem-resize>false</oem-resize>
         </oemconfig>
         <xsl:apply-templates mode="conv72to73"/>
      </type>
   </xsl:template>
```

When I use this input XML:

```xml
<type image="vmx"/>
```

I get this output:

```xml
<type image="oem">
  <oemconfig>
    <oem-resize>false</oem-resize>
  </oemconfig>
</type>
```

A `type` element with an `oemconfig` is not affected by this template rule.

Reply:
Thanks. I think your template is more robust and easier to read but it does not cover all cases. Here is the list of cases we are facing:

```xml
<type image="vmx"/>
```

should be transformed to

```xml
<type image="oem">
  <oemconfig>
    <oem-resize>false</oem-resize>
  </oemconfig>
</type>
```

This is covered nicely. Next there could also be the following:

```xml
<type image="vmx">
  <oemconfig>
    <oem-swap>true</oem-swap>
  </oemconfig>
</type>
```

It doesn't make sense as a vmx would not use the provided information but the schema out there would allow this setup and it should be transformed to:

```xml
<type image="oem">
  <oemconfig>
    <oem-resize>false</oem-resize>
    <oem-swap>true</oem-swap>
  </oemconfig>
</type>
```

I think this part is not covered by the current stylesheet

Reply:
and be aware oemconfig can contain many other elementes oem-swap was just taken as an example. So what we need is to keep the existing contents of oemconfig (except for oem-resize) and an explicit <oem-resize>false</oem-resize>

Reply:
Ok, I suggest an additional template rule like this:

```xml
<xsl:template match="type[@image = 'vmx']
                     [normalize-space(oemconfig/oem-swap) = 'true']" mode="conv72to73">
    <type image="oem">
         <oemconfig>
            <xsl:apply-templates select="oemconfig/*[not(self::oem-resize)]"
               mode="conv72to73"/>
            <oem-resize>false</oem-resize>
         </oemconfig>
    </type>
</xsl:template>
```

This would match for the second example and copy all elements from `oemconfig` except `oem-resize`.

-----

Additionally, maybe we should rewrite the first template a bit. At the moment, it looks like this:

```xml
<xsl:template match="type[@image = 'vmx'][not(oemconfig)]" mode="conv72to73">
```

This would match any `type` element without an `oemconfig` part. If you really want to match *only* an empty `<type image="vmx" />` you can rewrite the match like this:

```xml
<xsl:template match="type[@image = 'vmx'][not(*)]" mode="conv72to73">
```

Not sure if this would be better or worse. :wink:

Reply:
With all the info you provided I think we have a working version now